### PR TITLE
mobile_plugin: Mobile number process plugin added.

### PR DESCRIPTION
**Pull request by pranitjha**

### CODE REVIEW

File: src/Plugin/migrate/process/ApmMobileNumber.php [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+ * )
+ */
+class ApmMobileNumber extends ProcessPluginBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function transform($value, MigrateExecutableInterface $migrate_executable, Row $row, $destination_property) {
+    $mobile_data = $this->xml2array($value);
+    $util = \Drupal::service('mobile_number.util');
```
Comment (joshirohit100) [inline src/Plugin/migrate/process/ApmMobileNumber.php]: please inject it. Also add a check in constructor for the **mobile_number** module.

File: src/Plugin/migrate/process/ApmMobileNumber.php [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    // If mobile 'value' is not set throw error.
+    if (!isset($mobile_data['value'])) {
```
Comment (joshirohit100) [inline src/Plugin/migrate/process/ApmMobileNumber.php]: This is wrong. What if I have this value coming from the source in `my_mobile_number`?

File: src/Plugin/migrate/process/ApmMobileNumber.php [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+use Drupal\migrate\MigrateExecutableInterface;
+use Drupal\migrate\ProcessPluginBase;
+use Drupal\migrate\Row;
+use Drupal\mobile_number\MobileNumberUtilInterface;
+use Drupal\mobile_number\Element\MobileNumber;
+
+/**
+ * Migrates the Mobile number to mobile number field.
+ *
```
Comment (joshirohit100) [inline src/Plugin/migrate/process/ApmMobileNumber.php]: Please add some documentation about what it is expecting as source and what will it return. Also, what are the different config option it supports like `tfa`, `verified` etc and if those are not available, then whats the behavior.

File: src/Plugin/migrate/process/ApmMobileNumber.php [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  /**
+   * Function to convert XML to array.
+   *
+   * @param $xmlObject
+   *   XML object.
+   * @return array
+   *   Output data as array.
+   */
+  function xml2array ( $xmlObject) {
+    foreach ( (array) $xmlObject as $index => $node ) {
```
Comment (joshirohit100) [inline src/Plugin/migrate/process/ApmMobileNumber.php]: why we doing this? its not clear. can u pls provide some reference

**Comment by joshirohit100** [comment]
In the end what I think, that this should not be just drupal7 mobile_number migration field specific. It should be generic enough. For example, If I have a site (in any platform/technology) where I have mobile number stored in the DB and coming from the source like - `9971523642` or `+91-9971523642`. Then I should be able to migrate this into `mobile_number` field.  cc @amitgoyal

File: src/Plugin/migrate/process/ApmMobileNumber.php (hunk 4 above)
Comment (pranitjha) [reply src/Plugin/migrate/process/ApmMobileNumber.php]: To convert the simple XML object to array.

File: src/Plugin/migrate/process/ApmMobileNumber.php (hunk 3 above)
Comment (pranitjha) [reply src/Plugin/migrate/process/ApmMobileNumber.php]: Added additional comments to clarify.

File: src/Plugin/migrate/process/ApmMobileNumber.php (hunk 2 above)
Comment (pranitjha) [reply src/Plugin/migrate/process/ApmMobileNumber.php]: Corrected the code.

File: src/Plugin/migrate/process/ApmMobileNumber.php (hunk 1 above)
Comment (pranitjha) [reply src/Plugin/migrate/process/ApmMobileNumber.php]: Constructor added.

File: src/Plugin/migrate/process/ApmMobileNumber.php [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+ * Example of migration yml.
+ *
+ *  field_drupal_mobile_number:
+ *    plugin: apm_mobile_number
+ *    value: mobile_number #[Required]
+ *    country: x_country #[Optional if default_country is available]
+ *    local_number: x_local_number #[Optional]
+ *    verified: x_verified #[Optional, default is 0]
+ *    tfa: x_tfa #[Optional, default is 0]
+ *    default_country: IN #[Optional, if country is available for each record]
```
Comment (joshirohit100) [inline src/Plugin/migrate/process/ApmMobileNumber.php]: looks wrong comment.

File: src/Plugin/migrate/process/ApmMobileNumber.php [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+  /**
+   * {@inheritdoc}
+   */
+  public function transform($value, MigrateExecutableInterface $migrate_executable, Row $row, $destination_property) {
+    $processed_data = [];
+
+    // Converts simple XML object to array.
+    $mobile_data = $this->xml2array($value);
+
+    // If mobile 'value' is not skip the record and show error.
```
Comment (joshirohit100) [inline src/Plugin/migrate/process/ApmMobileNumber.php]: `value is not available`

File: src/Plugin/migrate/process/ApmMobileNumber.php [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+    if (!isset($mobile_data[$this->configuration['value']])) {
```
Comment (joshirohit100) [inline src/Plugin/migrate/process/ApmMobileNumber.php]: I think this should be done at constrcutor level as this should execute once if config in yml not available.

File: src/Plugin/migrate/process/ApmMobileNumber.php (hunk 4 above)
Comment (joshirohit100) [reply src/Plugin/migrate/process/ApmMobileNumber.php]: please add some example here as well to clear what this function accepts and returns.

**Comment by joshirohit100** [comment]
Looks fine to me but 2 feedback/comment.

File: src/Plugin/migrate/process/ApmMobileNumber.php (hunk 4 above)
Comment (pranitjha) [reply src/Plugin/migrate/process/ApmMobileNumber.php]: Function removed.

File: src/Plugin/migrate/process/ApmMobileNumber.php (hunk 5 above)
Comment (pranitjha) [reply src/Plugin/migrate/process/ApmMobileNumber.php]: Updated the comment.

File: src/Plugin/migrate/process/ApmMobileNumber.php (hunk 6 above)
Comment (pranitjha) [reply src/Plugin/migrate/process/ApmMobileNumber.php]: Updated the comment.

File: src/Plugin/migrate/process/ApmMobileNumber.php (hunk 7 above)
Comment (pranitjha) [reply src/Plugin/migrate/process/ApmMobileNumber.php]: Discussed and clarified.

File: modules/examples/apm_mobile_number_migrate/apm_mobile_number_migrate.info.yml [hunk 8]
```diff
@@ -0,0 +1,8 @@
+name: Acquia Platform Migration - Mobile Number Examples
```
Comment (joshirohit100) [inline modules/examples/apm_mobile_number_migrate/apm_mobile_number_migrate.info.yml]: I don't think u need new module. If u create a new module, then pls also import field config.

Comment (joshirohit100) [reply modules/examples/apm_mobile_number_migrate/apm_mobile_number_migrate.info.yml]: Discuss this with @nayna as how she doing this in her process plugin

File: src/Plugin/migrate/process/ApmMobileNumber.php [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+ *    <x_verified>0</x_verified>
+ *    <x_tfa>0</x_tfa>
+ *  </mobile>
+ *
+ * @MigrateProcessPlugin(
+ *   id = "apm_mobile_number"
+ * )
+ */
+class ApmMobileNumber extends ProcessPluginBase implements ContainerFactoryPluginInterface {
+  protected $mobile_number;
```
Comment (joshirohit100) [inline src/Plugin/migrate/process/ApmMobileNumber.php]: this is not as per coding standard. Check other plugins.

File: src/Plugin/migrate/process/ApmMobileNumber.php [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+
+  /**
+   * {@inheritdoc}
+   */
+  public function __construct(array $configuration, $pluginId, $pluginDefinition, ModuleHandlerInterface $moduleHandler) {
+    parent::__construct($configuration, $pluginId, $pluginDefinition);
+    if (!$moduleHandler->moduleExists('mobile_number')) {
```
Comment (joshirohit100) [inline src/Plugin/migrate/process/ApmMobileNumber.php]: Add some comment here.

File: src/Plugin/migrate/process/ApmMobileNumber.php [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+    }
+
+    // If mobile 'value' is not available skip the record and show error.
+    if (!isset($mobile_data[$field_mapping['value']])) {
+      throw new MigrateSkipRowException('The "value" must be set.');
+    }
+    $processed_data['value'] = $mobile_data[$field_mapping['value']];
+
+    // Set country as default country is value is not set.
+    if (empty($mobile_data[$field_mapping['country']])) {
```
Comment (joshirohit100) [inline src/Plugin/migrate/process/ApmMobileNumber.php]: Use php null coalescing instead if/else. See http://php.net/manual/en/migration70.new-features.php

File: src/Plugin/migrate/process/ApmMobileNumber.php [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+    }
+
+    // Generate mobile number that can be used to get callable and local number.
+    $mobile_number = $this->mobile_number->getMobileNumber($processed_data['value'], $processed_data['country']);
+
+    // Create mobile number value as per mobile_number module.
+    $processed_data['value'] = $this->mobile_number->getCallableNumber($mobile_number);
+
+    // Set local_number with help of mobile_number.
+    if (empty($mobile_data[$field_mapping['local_number']])) {
```
Comment (joshirohit100) [inline src/Plugin/migrate/process/ApmMobileNumber.php]: same here- use ternary operator.

File: src/Plugin/migrate/process/ApmMobileNumber.php [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+      $processed_data['local_number'] = $this->mobile_number->getLocalNumber($mobile_number);
+    }
+    else {
+      $processed_data['local_number'] = $mobile_data[$field_mapping['local_number']];
+    }
+
+    // If verified data is not available set default as 0.
+    if (empty($mobile_data[$field_mapping['verified']])) {
```
Comment (joshirohit100) [inline src/Plugin/migrate/process/ApmMobileNumber.php]: same here - ternary.

File: src/Plugin/migrate/process/ApmMobileNumber.php [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+      $processed_data['verified'] = 0;
+    }
+    else {
+      $processed_data['verified'] = $mobile_data[$field_mapping['verified']];
+    }
+
+    // If tfa data is not available set default as 0.
+    if (empty($mobile_data[$field_mapping['tfa']])) {
```
Comment (joshirohit100) [inline src/Plugin/migrate/process/ApmMobileNumber.php]: same here - ternary operator.

File: modules/examples/apm_mobile_number_migrate/apm_mobile_number_migrate.info.yml (hunk 8 above)
Comment (pranitjha) [reply modules/examples/apm_mobile_number_migrate/apm_mobile_number_migrate.info.yml]: Added field as configuration as discussed over slack.

File: src/Plugin/migrate/process/ApmMobileNumber.php (hunk 9 above)
Comment (pranitjha) [reply src/Plugin/migrate/process/ApmMobileNumber.php]: Changed the variable name as per standard.

File: src/Plugin/migrate/process/ApmMobileNumber.php (hunk 10 above)
Comment (pranitjha) [reply src/Plugin/migrate/process/ApmMobileNumber.php]: Comment added.

File: src/Plugin/migrate/process/ApmMobileNumber.php (hunk 11 above)
Comment (pranitjha) [reply src/Plugin/migrate/process/ApmMobileNumber.php]: used ternary operator.

File: src/Plugin/migrate/process/ApmMobileNumber.php (hunk 12 above)
Comment (pranitjha) [reply src/Plugin/migrate/process/ApmMobileNumber.php]: used ternary operator.

File: src/Plugin/migrate/process/ApmMobileNumber.php (hunk 13 above)
Comment (pranitjha) [reply src/Plugin/migrate/process/ApmMobileNumber.php]: used ternary operator.

File: src/Plugin/migrate/process/ApmMobileNumber.php (hunk 14 above)
Comment (pranitjha) [reply src/Plugin/migrate/process/ApmMobileNumber.php]: used ternary operator.

File: modules/examples/apm_mobile_number_migrate/config/install/field.field.user.user.field_mobile_number.yml [hunk 15]
```diff
@@ -0,0 +1,24 @@
+uuid: 6ad676e6-e66d-4d27-ba17-29eb7075b8f9
```
Comment (joshirohit100) [inline modules/examples/apm_mobile_number_migrate/config/install/field.field.user.user.field_mobile_number.yml]: this uuid not required

File: modules/examples/apm_mobile_number_migrate/config/install/field.storage.user.field_mobile_number.yml [hunk 16]
```diff
@@ -0,0 +1,20 @@
+uuid: 1a43e300-786a-48cc-9115-5900c2a44e45
```
Comment (joshirohit100) [inline modules/examples/apm_mobile_number_migrate/config/install/field.storage.user.field_mobile_number.yml]: same here.

**Comment by joshirohit100** [comment]
Looks fine to me excpet two minor uuid comments.

cc @amitgoyal for review and merge.

File: modules/examples/apm_mobile_number_migrate/apm_mobile_number_migrate.info.yml [hunk 17]
```diff
@@ -0,0 +1,8 @@
...
+type: module
+description: 'XML to Drupal 8 migration examples.'
```
Comment (amitgoyal) [inline modules/examples/apm_mobile_number_migrate/apm_mobile_number_migrate.info.yml]: Wrong description?

File: modules/examples/apm_mobile_number_migrate/README.md [hunk 18]
```diff
@@ -0,0 +1,10 @@
+# Mobile Number migration
+
+This is a sample module to migrate data to Mobile Number field.
```
Comment (amitgoyal) [inline modules/examples/apm_mobile_number_migrate/README.md]: Can we add bit more description indicating the use case of this module / process plugin.

**Pull request merged**